modules/qualitycontrol/SendVideoController.js

```diff
@@ -50,26 +51,52 @@ export class SendVideoController {
         mediaSession.addListener(
             MediaSessionEvents.REMOTE_VIDEO_CONSTRAINTS_CHANGED,
             session => {
-                if (session === this.conference._getActiveMediaSession()) {
+                if (session === this._conference.getActiveMediaSession()) {
                     this._propagateSendMaxFrameHeight();
                 }
             });
     }
 
+    /**
+     * Propagates the video constraints if they have changed.
+     *
+     * @param {Object} videoConstraints - The sender video constraints received from the bridge.
+     */
+    _onSenderConstraintsReceived(videoConstraints) {
+        if (FeatureFlags.isSourceNameSignalingEnabled()) {
+            const { idealHeight, sourceName } = videoConstraints;
+            const localVideoTracks = this._conference.getLocalVideoTracks() ?? [];
+
+            for (const track of localVideoTracks) {
+                // Propagate the sender constraint only if it has changed.
+                if (track.getSourceName() === sourceName
+                    && (!this._sourceSenderConstraints.has(sourceName)
+                    || this._sourceSenderConstraints.get(sourceName) !== idealHeight)) {
+                    this._sourceSenderConstraints.set(sourceName, idealHeight);
+                    this._propagateSendMaxFrameHeight(sourceName);
+                }
+            }
+        } else if (this._senderVideoConstraints?.idealHeight !== videoConstraints.idealHeight) {
+            this._senderVideoConstraints = videoConstraints;
+            this._propagateSendMaxFrameHeight();
+        }
+    }
+
     /**
      * Figures out the send video constraint as specified by {@link selectSendMaxFrameHeight} and sets it on all media
      * sessions for the reasons mentioned in this class description.
      *
+     * @param {string} sourceName - The source for which sender constraints have changed.
      * @returns {Promise<void[]>}
      * @private
      */
-    _propagateSendMaxFrameHeight() {
-        const sendMaxFrameHeight = this.selectSendMaxFrameHeight();
+    _propagateSendMaxFrameHeight(sourceName = null) {
```

Comment:
Hmm, that would be lot of code duplication. I can add more debugging messages there to make it easier to debug issues. Since the source name is only passed to these methods when the flag is enabled, do you still think that bugs might creep in ?

Reply:
yes, I think the code should fail fast when there's no source name in the new flow, ideally by throwing an error, because "source name should always be there"